### PR TITLE
Integration of a videoconferencing functionality using Jitsi.

### PR DESCRIPTION
**Pull request by Orodan**
Hi,

please find in this pull request the work done at the University of La Rochelle to integrate a videoconferencing system in OAE using the Jitsi technology. Jitsi is a open source and free to use application which allow people to create rooms in which they can chat and interact together with a video and audio stream for example. You can find more informations about Jitsi there : https://jitsi.org/.

You will find most of the work done in the oae-jitsi-widgets folder, which contains all the widgets added for the integration of the videoconferencing functionality. I also added a new page in the ui to display a meeting and a new library to list and search them. Finally the project contains the last updates of the oaeproject/3akai-ux master branch.

One crucial point : to allow video and audio stream, you have to be connected in OAE in https (it is a browser restriction). I wanted to precise it because my repo does not contain the nginx conf I used for my development.

You can find the pull request for the Hilary project here : https://github.com/oaeproject/Hilary/pull/1299 and the pull request for the oae-rest project here : https://github.com/oaeproject/oae-rest/pull/33

Feel free to contact me if you have any questions or remarks.

Best regards

Jimmy

PS : because it was simpler to work this way, most of the work was not done in this branch but in two others repositories :

https://github.com/Orodan/3akai-ux-jitsi-fork for the 3akai-ux project.
https://github.com/Orodan/oae-jitsi-widgets for the oae-jitsi-widget folder.

### CODE REVIEW

File: nginx/nginx.conf [hunk 1]
```diff
@@ -1,4 +1,4 @@
-user                    <%= nginxConf.NGINX_USER %> <%= nginxConf.NGINX_GROUP %>;
+user                    jimmykasprzak staff;
```
Comment (sallakarppinen) [inline nginx/nginx.conf]: Hello, I think this config file was included in the pull request by accident, could you take it out please? :)

Comment (Orodan) [inline nginx/nginx.conf]: Indeed, it's done :)

Comment (sallakarppinen) [inline nginx/nginx.conf]: Thanks!

File: nginx/nginx.conf [hunk 2]
```diff
@@ -632,4 +632,4 @@ http {
     }
 
     include <%= nginxConf.UX_HOME %>/nginx/mime.conf;
-}
+}
```
Comment (sallakarppinen) [inline nginx/nginx.conf]: Files should end with a newline.

File: node_modules/oae-core/deleteresource/js/deleteresource.js [hunk 3]
```diff
@@ -70,6 +72,8 @@ define(['jquery', 'oae.core'], function ($, oae) {
          * based on the resource type
          */
         var setUpDeleteResource = function() {
+            console.log('setUpDeleteResource');
```
Comment (sallakarppinen) [inline node_modules/oae-core/deleteresource/js/deleteresource.js]: Remove console.log statements.

File: node_modules/oae-jitsi-widgets/aboutmeeting-jitsi/bundles/default.properties [hunk 4]
```diff
@@ -0,0 +1,2 @@
+ADDED_BY_DISPLAYNAME = Added by ${displayNameLink}
+ADDED_COLON = Added:
```
Comment (sallakarppinen) [inline node_modules/oae-jitsi-widgets/aboutmeeting-jitsi/bundles/default.properties]: Files should end with newline.

**Comment by sallakarppinen** [comment]
1. All files should end with a newline;
2. Remove binary file .DS_Store.

File: node_modules/oae-core/deleteresource/js/deleteresource.js (hunk 3 above)
Comment (brecke) [inline node_modules/oae-core/deleteresource/js/deleteresource.js]: Fixed in 51ad3db

File: nginx/nginx.conf (hunk 2 above)
Comment (brecke) [inline nginx/nginx.conf]: Fixed in 51ad3db

**Comment by brecke** [comment]
Superseded by PR [#4122](https://github.com/oaeproject/3akai-ux/pull/4122)

Some comments above have been fixed in 0649988 and 51ad3db

File: node_modules/oae-jitsi-widgets/meeting-jitsi/js/meeting-jitsi.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            if (widgetData.contactList === true) toggleContactList();
+        };
+
+        /**
+         * Render the meeting room.
+         */
+        var renderMeeting = function () {
+
+            var domain = oae.api.config.getValue('oae-jitsi', 'server', 'host');
+            var room = "JitsiMeetAPIExample";
```
Comment (sallakarppinen) [inline node_modules/oae-jitsi-widgets/meeting-jitsi/js/meeting-jitsi.js]: We need to change the room name - could use the `meetingId` to make sure we always send people to the correct room?

Comment (brecke) [inline node_modules/oae-jitsi-widgets/meeting-jitsi/js/meeting-jitsi.js]: Fixed in 77330a1

File: ui/js/group.js [hunk 6]
```diff
@@ -160,6 +160,12 @@ require(['jquery', 'oae.core'], function($, oae) {
                         'title': oae.api.i18n.translate('__MSG__DISCUSSION__'),
                         'closeNav': true,
                         'class': 'oae-trigger-creatediscussion'
+                    },
+                    {
```
Comment (sallakarppinen) [inline ui/js/group.js]: It seems that this remains visible on the group pages even when no Jitsi server is set. Meetings should be disabled across board when there's no server in the config.

File: ui/js/group.js [hunk 7]
```diff
@@ -272,7 +278,28 @@ require(['jquery', 'oae.core'], function($, oae) {
                     ]
                 }
             ]
-        });
+        },
+        {
```
Comment (sallakarppinen) [inline ui/js/group.js]: See above; these options should not be visible when no Jitsi server is set.

File: ui/js/meeting-jitsi.js [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            'defaultRole': 'member',
+            'roles': [
+                {'id': 'member', 'name': oae.api.i18n.translate('__MSG__CAN_VIEW__')},
+                {'id': 'manager', 'name': oae.api.i18n.translate('__MSG__CAN_MANAGE__')}
+            ],
+            'api': {
+                'getMembersURL': '/api/meeting-jitsi/'+ meetingProfile.id + '/members',
+                'setMembers': oae.api.meetingJitsi.updateMembers,
+                'getInvitations': oae.api.meetingJitsi.getInvitations,
+                'resendInvitation': oae.api.meetingJitsi.resendInvitation, // to do
```
Comment (sallakarppinen) [inline ui/js/meeting-jitsi.js]: To do? We need to check whether this has been completed.

File: ui/js/meeting-jitsi.js [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        oae.api.util.template().render($('#meeting-jitsi-clip-template'), {
+            'meeting': meetingProfile,
+            'displayOptions': {
+                'addLink:': false
+            }
+        }, $('#meeting-jitsi-clip-container'));
+    };
+
+    /**
+     * TODO
```
Comment (sallakarppinen) [inline ui/js/meeting-jitsi.js]: Same as above.

File: node_modules/oae-jitsi-widgets/meeting-jitsi/js/meeting-jitsi.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+define(['jquery', 'oae.core', 'postis', 'JitsiMeetExternalAPI'], function ($, oae, postis, JitsiMeetExternalAPI) {
+
+    return function (uid, showSettings, widgetData) {
+
+        // getting meetingId to use as room id
+        var meetingId = _.last(widgetData.id.split(':'));
```
Comment (davidoae) [inline node_modules/oae-jitsi-widgets/meeting-jitsi/js/meeting-jitsi.js]: I am sure it'll not be enough.

File: shared/vendor/js/JitsiMeetExternalAPI.js [hunk 11]
```diff
@@ -0,0 +1,411 @@
+/*!
+ * Copyright 2014 Apereo Foundation (AF) Licensed under the
```
Comment (sallakarppinen) [inline shared/vendor/js/JitsiMeetExternalAPI.js]: I don't think we can put our license here, since this is adapted from jitsi-meet project?

Comment (brecke) [inline shared/vendor/js/JitsiMeetExternalAPI.js]: My bad, I didn't think it through

File: node_modules/oae-jitsi-widgets/meeting-jitsi/js/meeting-jitsi.js (hunk 10 above)
Comment (sallakarppinen) [inline node_modules/oae-jitsi-widgets/meeting-jitsi/js/meeting-jitsi.js]: I think so as well. Will get back to it tomorrow.

**Comment by davidoae** [comment]
They use Apache version 2 license, which our educational license is
compatible with.
Therefore when we adapt jitsi-meet we can indeed re-license the adapted
code to the Educational Community License 2.0.

D

**Pull request merged**